merlin-server/sql/merlin/hasura_functions.sql

```diff
@@ -231,3 +231,91 @@ begin
+  begin
+    if not exists(select id from public.activity_directive where (id, plan_id) = (_activity_id, _plan_id)) then
+      raise exception 'Activity Directive % does not exist in Plan %', _activity_id, _plan_id;
+    end if;
+
+    return query
+      with updated as (
+        select public.anchor_direct_descendents_to_plan(_activity_id := _activity_id, _plan_id := _plan_id)
+      )
+      select updated.*, 'updated'
+        from updated;
+
+    return query
+      with deleted as (
+        delete from activity_directive where (id, plan_id) = (_activity_id, _plan_id) returning *
+      )
+      select (deleted.id, deleted.plan_id, deleted.name, deleted.tags, deleted.source_scheduling_goal_id,
+              deleted.created_at, deleted.last_modified_at, deleted.start_offset, deleted.type, deleted.arguments,
+              deleted.last_modified_arguments_at, deleted.metadata, deleted.anchor_id, deleted.anchored_to_start)::activity_directive, 'deleted' from deleted;
+  end
+$$;
+
+create function hasura_functions.delete_activity_by_pk_reanchor_to_anchor(_activity_id int, _plan_id int)
+  returns setof hasura_functions.delete_anchor_return_value
+  strict
+  language plpgsql as $$
+begin
+  if not exists(select id from public.activity_directive where (id, plan_id) = (_activity_id, _plan_id)) then
+    raise exception 'Activity Directive % does not exist in Plan %', _activity_id, _plan_id;
+  end if;
+
+    return query
+      with updated as (
+        select public.anchor_direct_descendents_to_ascendant_anchor(_activity_id := _activity_id, _plan_id := _plan_id)
+      )
+      select updated.*, 'updated'
+        from updated;
+    return query
+      with deleted as (
+        delete from activity_directive where (id, plan_id) = (_activity_id, _plan_id) returning *
+      )
+      select (deleted.id, deleted.plan_id, deleted.name, deleted.tags, deleted.source_scheduling_goal_id,
+              deleted.created_at, deleted.last_modified_at, deleted.start_offset, deleted.type, deleted.arguments,
+              deleted.last_modified_arguments_at, deleted.metadata, deleted.anchor_id, deleted.anchored_to_start)::activity_directive, 'deleted' from deleted;
+end
+$$;
+
+create function hasura_functions.delete_activity_by_pk_delete_chain(_activity_id int, _plan_id int)
+  returns setof hasura_functions.delete_anchor_return_value
+  strict
+  language plpgsql as $$
+begin
+  if not exists(select id from public.activity_directive where (id, plan_id) = (_activity_id, _plan_id)) then
+    raise exception 'Activity Directive % does not exist in Plan %', _activity_id, _plan_id;
+  end if;
+
+  return query
+    with recursive descendents(activity_id, p_id) as (
+    select _activity_id, _plan_id
+    from activity_directive ad
+    where (ad.id, ad.plan_id) = (_activity_id, _plan_id)
+    union
+        select ad.id, ad.plan_id
+        from activity_directive ad, descendents d
+        where (ad.anchor_id, ad.plan_id) = (d.activity_id, d.p_id)
```

Comment:
Not sure what the best indentation is for this kind of query - here's one that seems okay 😅

```suggestion
    with recursive descendents(activity_id, p_id) as (
      select _activity_id, _plan_id
        from activity_directive ad
        where (ad.id, ad.plan_id) = (_activity_id, _plan_id)
    union
      select ad.id, ad.plan_id
        from activity_directive ad, descendents d
        where (ad.anchor_id, ad.plan_id) = (d.activity_id, d.p_id)
```
The `from` and `where` can be aligned with the select if you prefer - I'm mostly looking for indentation relative to the `with` and `union` parts of the recursive query

Reply:
Thanks. I really need to disable auto-indent for SQL files 🤦‍♀️